src/main/java/com/google/devtools/build/lib/analysis/constraints/RuleContextConstraintSemantics.java

```diff
@@ -808,4 +825,147 @@ private static void addSelectValuesToSet(BuildType.Selector<?> select, final Set
       type.visitLabels(visitor, value, /*context=*/ null);
     }
   }
+
+  /**
+   * Creates an incompatible {@link ConfiguredTarget} if the corresponding rule is incompatible.
+   *
+   * <p>Returns null if the target is not incompatible.
+   *
+   * <p>"Incompatible" here means either:
+   *
+   * <ol>
+   *   <li>the corresponding <code>target_compatible_with</code> list contains a constraint that the
+   *       current platform doesn't satisfy, or
+   *   <li>a transitive dependency is incompatible.
+   * </ol>
+   *
+   * @param ruleContext analysis context for the rule
+   * @param prerequisiteMap the dependencies of the rule
+   * @throws ActionConflictException if the underlying {@link RuleConfiguredTargetBuilder}
+   *     encounters a problem when assembling a dummy action for the incompatible {@link
+   *     ConfiguredTarget}.
+   */
+  @Nullable
+  public static ConfiguredTarget incompatibleConfiguredTarget(
+      RuleContext ruleContext,
+      OrderedSetMultimap<DependencyKind, ConfiguredTargetAndData> prerequisiteMap)
+      throws ActionConflictException {
+    if ("toolchain".equals(ruleContext.getRule().getRuleClass())) {
```

Comment:
Explicitly exempting the "toolchain" rule makes sense (it uses target_compatible_with very differently), but could this just re-use the existing RuleClass.useToolchainResolution() check instead? Anything not using toolchain resolution should also not check target compatibility.

Reply:
Good call, done.